wallet/chain/p/builder/builder.go

```diff
@@ -1194,3 +1493,173 @@ func (b *builder) initCtx(tx txs.UnsignedTx) error {
+	weights  feecomponent.Dimensions
+	gasPrice feecomponent.GasPrice
+
+	amountsToBurn  map[ids.ID]uint64
+	amountsToStake map[ids.ID]uint64
+	complexity     feecomponent.Dimensions
+
+	inputs        []*avax.TransferableInput
+	changeOutputs []*avax.TransferableOutput
+	stakeOutputs  []*avax.TransferableOutput
+}
+
+func (s *spendHelper) addInput(input *avax.TransferableInput) error {
+	newInputComplexity, err := txfee.InputComplexity(input)
+	if err != nil {
+		return err
+	}
+	s.complexity, err = s.complexity.Add(newInputComplexity)
+	if err != nil {
+		return err
+	}
+
+	s.inputs = append(s.inputs, input)
+	return nil
+}
+
+func (s *spendHelper) addChangeOutput(output *avax.TransferableOutput) error {
+	s.changeOutputs = append(s.changeOutputs, output)
+	return s.addOutputComplexity(output)
+}
+
+func (s *spendHelper) addStakedOutput(output *avax.TransferableOutput) error {
+	s.stakeOutputs = append(s.stakeOutputs, output)
+	return s.addOutputComplexity(output)
+}
+
+func (s *spendHelper) addOutputComplexity(output *avax.TransferableOutput) error {
+	newOutputComplexity, err := txfee.OutputComplexity(output)
+	if err != nil {
+		return err
+	}
+	s.complexity, err = s.complexity.Add(newOutputComplexity)
+	return err
+}
+
+func (s *spendHelper) shouldConsumeLockedAsset(assetID ids.ID) bool {
+	remainingAmountToStake := s.amountsToStake[assetID]
+	return remainingAmountToStake != 0
+}
+
+func (s *spendHelper) shouldConsumeAsset(assetID ids.ID) bool {
+	remainingAmountToBurn := s.amountsToBurn[assetID]
+	remainingAmountToStake := s.amountsToStake[assetID]
+	return remainingAmountToBurn != 0 || remainingAmountToStake != 0
+}
+
+func (s *spendHelper) consumeLockedAsset(assetID ids.ID, amount uint64) uint64 {
+	remainingAmountToStake := s.amountsToStake[assetID]
+	// Stake any value that should be staked
+	amountToStake := min(
+		remainingAmountToStake, // Amount we still need to stake
+		amount,                 // Amount available to stake
+	)
+	s.amountsToStake[assetID] -= amountToStake
+	return amount - amountToStake
+}
+
+func (s *spendHelper) consumeAsset(assetID ids.ID, amount uint64) uint64 {
+	remainingAmountToBurn := s.amountsToBurn[assetID]
+	remainingAmountToStake := s.amountsToStake[assetID]
+
+	// Burn any value that should be burned
+	amountToBurn := min(
+		remainingAmountToBurn, // Amount we still need to burn
+		amount,                // Amount available to burn
+	)
+	s.amountsToBurn[assetID] -= amountToBurn
+
+	amountAvailableToStake := amount - amountToBurn
+	// Burn any value that should be burned
+	amountToStake := min(
+		remainingAmountToStake, // Amount we still need to stake
+		amountAvailableToStake, // Amount available to stake
+	)
+	s.amountsToStake[assetID] -= amountToStake
+	return amountAvailableToStake - amountToStake
+}
+
+func (s *spendHelper) calculateFee() (uint64, error) {
+	gas, err := s.complexity.ToGas(s.weights)
+	if err != nil {
+		return 0, err
+	}
+	return math.Mul(uint64(s.gasPrice), uint64(gas))
```

Comment:
I really think we should reuse the proper fee Calculator here. I think this is a dangerour code duplication in case we change the way we map complexity to gas

Reply:
I modified this method so that we only convert `gas` and `gasPrice` to `nAVAX` in one location. lmk if that addresses your concern.